File: airbyte-commons-worker/src/test/java/io/airbyte/workers/internal/SimpleAirbyteDestination.java
/*
 * Copyright (c) 2020-2025 Airbyte, Inc., all rights reserved.
 */

package io.airbyte.workers.internal;

import io.airbyte.config.WorkerDestinationConfig;
import io.airbyte.protocol.models.AirbyteMessage;
import io.airbyte.protocol.models.AirbyteMessage.Type;
import java.nio.file.Path;
import java.util.Optional;
import java.util.concurrent.BlockingQueue;
import java.util.concurrent.LinkedBlockingQueue;

/**
 * Simple in memory implementation of an AirbyteDestination for testing purpose.
 */
public class SimpleAirbyteDestination implements AirbyteDestination {

  private final BlockingQueue<AirbyteMessage> messages = new LinkedBlockingQueue<>();
  private volatile boolean isFinished = false;

  @Override
  public void start(WorkerDestinationConfig destinationConfig, Path jobRoot) throws Exception {}

  @Override
  public void accept(AirbyteMessage message) throws Exception {
    if (message.getType() == Type.STATE) {
      messages.put(message);
    }
  }

  @Override
  public void notifyEndOfInput() throws Exception {
    isFinished = true;
  }

  @Override
  public boolean isFinished() {
    return isFinished && messages.isEmpty();
  }

  @Override
  public int getExitValue() {
    return 0;
  }

  @Override
  public Optional<AirbyteMessage> attemptRead() {
    if (messages.isEmpty() && isFinished) {
      return Optional.empty();
    }
    return Optional.ofNullable(messages.poll());
  }

  @Override
  public void close() throws Exception {

  }

  @Override
  public void cancel() throws Exception {

  }

}

File: airbyte-container-orchestrator/src/test/kotlin/io/airbyte/container/orchestrator/worker/SimpleAirbyteDestination.kt
/*
 * Copyright (c) 2020-2025 Airbyte, Inc., all rights reserved.
 */

package io.airbyte.container.orchestrator.worker

import io.airbyte.config.WorkerDestinationConfig
import io.airbyte.protocol.models.v0.AirbyteMessage
import io.airbyte.workers.internal.AirbyteDestination
import java.nio.file.Path
import java.util.Optional
import java.util.concurrent.BlockingQueue
import java.util.concurrent.LinkedBlockingQueue
import kotlin.concurrent.Volatile

/**
 * Simple in-memory implementation of an AirbyteDestination for testing purpose.
 */
class SimpleAirbyteDestination : AirbyteDestination {
  private val messages: BlockingQueue<AirbyteMessage> = LinkedBlockingQueue()

  @Volatile
  private var isFinished = false

  @Throws(Exception::class)
  override fun start(
    destinationConfig: WorkerDestinationConfig,
    jobRoot: Path,
  ) {
  }

  @Throws(Exception::class)
  override fun accept(message: AirbyteMessage) {
    if (message.getType() == AirbyteMessage.Type.STATE) {
      messages.put(message)
    }
  }

  @Throws(Exception::class)
  override fun notifyEndOfInput() {
    isFinished = true
  }

  override fun isFinished(): Boolean = isFinished && messages.isEmpty()

  override fun getExitValue(): Int = 0

  override fun attemptRead(): Optional<AirbyteMessage> {
    if (messages.isEmpty() && isFinished) {
      return Optional.empty<AirbyteMessage>()
    }
    return Optional.ofNullable<AirbyteMessage>(messages.poll())
  }

  @Throws(Exception::class)
  override fun close() {
  }

  @Throws(Exception::class)
  override fun cancel() {
  }
}
